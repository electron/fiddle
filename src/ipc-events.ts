export enum IpcEvents {
  OPEN_SETTINGS = 'OPEN_SETTINGS',
  LOAD_GIST_REQUEST = 'LOAD_GIST_REQUEST',
  LOAD_ELECTRON_EXAMPLE_REQUEST = 'LOAD_ELECTRON_EXAMPLE_REQUEST',
  FIDDLE_RUN = 'FIDDLE_RUN',
  FIDDLE_PACKAGE = 'FIDDLE_PACKAGE',
  FIDDLE_MAKE = 'FIDDLE_MAKE',
  MONACO_EXECUTE_COMMAND = 'MONACO_EXECUTE_COMMAND',
  MONACO_TOGGLE_OPTION = 'MONACO_TOGGLE_OPTION',
  FS_NEW_FIDDLE = 'FS_NEW_FIDDLE',
  FS_NEW_TEST = 'FS_NEW_TEST',
  FS_OPEN_FIDDLE = 'FS_OPEN_FIDDLE',
  FS_OPEN_TEMPLATE = 'FS_OPEN_TEMPLATE',
  FS_SAVE_FIDDLE = 'FS_SAVE_FIDDLE',
  FS_SAVE_FIDDLE_AS = 'FS_SAVE_FIDDLE_AS',
  FS_SAVE_FIDDLE_GIST = 'FS_SAVE_FIDDLE_GIST',
  FS_SAVE_FIDDLE_FORGE = 'FS_SAVE_FIDDLE_FORGE',
  FS_SAVE_FIDDLE_DIALOG = 'FS_SAVE_FIDDLE_DIALOG',
  FS_SAVE_FIDDLE_ERROR = 'FS_SAVE_FIDDLE_ERROR',
  SHOW_WARNING_DIALOG = 'SHOW_WARNING_DIALOG',
  SHOW_CONFIRMATION_DIALOG = 'SHOW_CONFIRMATION_DIALOG',
  SHOW_WELCOME_TOUR = 'SHOW_WELCOME_TOUR',
  CLEAR_CONSOLE = 'CLEAR_CONSOLE',
  LOAD_LOCAL_VERSION_FOLDER = 'LOAD_LOCAL_VERSION_FOLDER',
  SHOW_LOCAL_VERSION_FOLDER_DIALOG = 'SHOW_LOCAL_VERSION_FOLDER_DIALOG',
  BISECT_COMMANDS_TOGGLE = 'BISECT_COMMANDS_TOGGLE',
  SHOW_INACTIVE = 'SHOW_INACTIVE',
  BEFORE_QUIT = 'BEFORE_QUIT',
  CONFIRM_QUIT = 'CONFIRM_QUIT',
  GET_APP_PATHS = 'GET_APP_PATHS',
  SELECT_ALL_IN_EDITOR = 'SELECT_ALL_IN_EDITOR',
  BLOCK_ACCELERATORS = 'BLOCK_ACCELERATORS',
  SET_SHOW_ME_TEMPLATE = 'SET_SHOW_ME_TEMPLATE',
  CLICK_TITLEBAR_MAC = 'CLICK_TITLEBAR_MAC',
  TASK_BISECT = 'TASK_BISECT',
  TASK_TEST = 'TASK_TEST',
  TASK_DONE = 'TASK_DONE',
  OUTPUT_ENTRY = 'OUTPUT_ENTRY',
  RELOAD_WINDOW = 'RELOAD_WINDOW',
}

export const ipcMainEvents = [
  IpcEvents.FS_SAVE_FIDDLE_DIALOG,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.SHOW_WARNING_DIALOG,
  IpcEvents.SHOW_CONFIRMATION_DIALOG,
  IpcEvents.SHOW_LOCAL_VERSION_FOLDER_DIALOG,
  IpcEvents.SHOW_INACTIVE,
  IpcEvents.CONFIRM_QUIT,
  IpcEvents.SET_SHOW_ME_TEMPLATE,
  IpcEvents.BLOCK_ACCELERATORS,
  IpcEvents.CLICK_TITLEBAR_MAC,
  IpcEvents.OUTPUT_ENTRY,
  IpcEvents.TASK_DONE,
  IpcEvents.RELOAD_WINDOW,
];

export const ipcRendererEvents = [
  IpcEvents.OPEN_SETTINGS,
  IpcEvents.LOAD_GIST_REQUEST,
  IpcEvents.LOAD_ELECTRON_EXAMPLE_REQUEST,
  IpcEvents.FIDDLE_RUN,
  IpcEvents.FIDDLE_PACKAGE,
  IpcEvents.FIDDLE_MAKE,
  IpcEvents.MONACO_EXECUTE_COMMAND,
  IpcEvents.MONACO_TOGGLE_OPTION,
  IpcEvents.FS_NEW_FIDDLE,
  IpcEvents.FS_NEW_TEST,
  IpcEvents.FS_OPEN_FIDDLE,
  IpcEvents.FS_OPEN_TEMPLATE,
  IpcEvents.FS_SAVE_FIDDLE,
  IpcEvents.FS_SAVE_FIDDLE_AS,
  IpcEvents.FS_SAVE_FIDDLE_FORGE,
  IpcEvents.FS_SAVE_FIDDLE_GIST,
  IpcEvents.SHOW_WELCOME_TOUR,
  IpcEvents.CLEAR_CONSOLE,
  IpcEvents.LOAD_LOCAL_VERSION_FOLDER,
  IpcEvents.BISECT_COMMANDS_TOGGLE,
  IpcEvents.BEFORE_QUIT,
  IpcEvents.GET_APP_PATHS,
  IpcEvents.SELECT_ALL_IN_EDITOR,
  IpcEvents.SET_SHOW_ME_TEMPLATE,
  IpcEvents.TASK_BISECT,
  IpcEvents.TASK_TEST,
];

export const WEBCONTENTS_READY_FOR_IPC_SIGNAL =
  'WEBCONTENTS_READY_FOR_IPC_SIGNAL';
