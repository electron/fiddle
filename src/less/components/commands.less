@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-3;
  -webkit-app-region: drag;
}

.disabled-menu-tooltip {
  display: flex;
  width: 100%;

  .bp3-popover-target {
    width: 100%;
  }
}

.commands.is-mac {
  padding:
    10px
    calc(100vw - (env(titlebar-area-x, 0) + env(titlebar-area-width, 0)))
    10px
    env(titlebar-area-x, 0);
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding:
    10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
    white-space: nowrap;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  fieldset {
    border: 0;
    padding: 0;
    margin: 0;
    min-width: 0;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      .bp3-button-group {
        margin-left: 5px;
      }
    }
  }

  .bp3-control-group {
    margin-right: 0.5rem;
  }

  .address-bar {
    width: 420px;
    transition: width 0.4s;

    input,
    span {
      -webkit-app-region: no-drag;
    }

    &.empty {
      width: 300px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }

  & > .title {
    align-self: stretch;
    flex-shrink: 0;

    flex-flow: column nowrap;
    justify-content: center;
    user-select: none;

    margin: 0 10px;
  }
}
