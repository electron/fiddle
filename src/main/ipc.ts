import { EventEmitter } from 'events';

import { ipcMain } from 'electron';

import {
  IpcEvents,
  WEBCONTENTS_READY_FOR_IPC_SIGNAL,
  ipcMainEvents,
} from '../ipc-events';
import { getOrCreateMainWindow } from './windows';

/**
 * The main purpose of this class is to be the central
 * gathering place for IPC calls the main process sends
 * or listens to.
 *
 * @class IpcManager
 * @extends {EventEmitter}
 */
class IpcMainManager extends EventEmitter {
  public readyWebContents = new WeakSet<Electron.WebContents>();
  private messageQueue = new WeakMap<
    Electron.WebContents,
    Array<[IpcEvents, Array<any> | undefined]>
  >();

  constructor() {
    super();

    ipcMainEvents.forEach((name) => {
      ipcMain.removeAllListeners(name);
      ipcMain.on(name, (...args: Array<any>) => this.emit(name, ...args));
    });

    ipcMain.on(
      WEBCONTENTS_READY_FOR_IPC_SIGNAL,
      (event: Electron.IpcMainEvent) => {
        this.readyWebContents.add(event.sender);

        const queue = this.messageQueue.get(event.sender);
        this.messageQueue.delete(event.sender);
        if (!queue) return;
        for (const item of queue) {
          this.send(item[0], item[1], event.sender);
        }
      },
    );
  }

  /**
   * Send an IPC message to an instance of Electron.WebContents.
   * If none is specified, we'll automatically go with the main window.
   *
   * @param {IpcEvents} channel
   * @param {Array<any>} [args]
   * @param {Electron.WebContents} [target]
   */
  public send(
    channel: IpcEvents,
    args?: Array<any>,
    target?: Electron.WebContents,
  ) {
    const _target = target || getOrCreateMainWindow().webContents;
    const _args = args || [];
    if (!this.readyWebContents.has(_target)) {
      const existing = this.messageQueue.get(_target) || [];
      this.messageQueue.set(_target, [...existing, [channel, args]]);
      return;
    }

    _target.isDestroyed() || _target.send(channel, ..._args);
  }

  public handle(
    channel: IpcEvents,
    listener: (event: Electron.IpcMainInvokeEvent, ...args: any[]) => any,
  ) {
    // there can be only one, so remove previous one first
    ipcMain.removeHandler(channel);
    ipcMain.handle(channel, listener);
  }

  public handleOnce(
    channel: IpcEvents,
    listener: (event: Electron.IpcMainInvokeEvent, ...args: any[]) => any,
  ) {
    ipcMain.handleOnce(channel, listener);
  }
}

export const ipcMainManager = new IpcMainManager();
