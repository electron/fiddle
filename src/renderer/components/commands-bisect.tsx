import * as React from 'react';

import { Button } from '@blueprintjs/core';
import { InstallState } from '@vertedinde/fiddle-core';
import { observer } from 'mobx-react';

import { AppState } from '../state';

interface BisectHandlerProps {
  appState: AppState;
}

export const BisectHandler = observer(
  class BisectHandler extends React.Component<BisectHandlerProps> {
    constructor(props: BisectHandlerProps) {
      super(props);

      this.continueBisect = this.continueBisect.bind(this);
      this.terminateBisect = this.terminateBisect.bind(this);
    }

    private continueBisect(isGood: boolean) {
      window.ElectronFiddle.app.runner.stop();

      const { appState } = this.props;
      const response = appState.Bisector!.continue(isGood);

      if (Array.isArray(response)) {
        this.terminateBisect();

        const [minRev, maxRev] = response;
        const [minVer, maxVer] = [minRev.version, maxRev.version];
        const label = (
          <>
            Bisect complete. Check the range{' '}
            <a
              target="_blank"
              rel="noreferrer"
              href={`https://github.com/electron/electron/compare/v${minVer}...v${maxVer}`}
            >
              {minVer}...{maxVer}
            </a>
            .
          </>
        );

        appState.pushOutput(`[BISECT] Complete: ${minVer}...${maxVer}`);
        appState.showInfoDialog(label);
      } else {
        appState.setVersion(response.version);
      }
    }

    private terminateBisect() {
      const { appState } = this.props;
      appState.Bisector = undefined;
    }

    public render() {
      const { appState } = this.props;
      if (!!appState.Bisector) {
        const isDownloading =
          appState.currentElectronVersion.state === InstallState.downloading;
        return (
          <>
            <Button
              icon={'thumbs-up'}
              aria-label={'Mark commit as good'}
              onClick={() => this.continueBisect(true)}
              disabled={isDownloading}
            />
            <Button
              icon={'thumbs-down'}
              aria-label={'Mark commit as bad'}
              onClick={() => this.continueBisect(false)}
              disabled={isDownloading}
            />
            <Button
              aria-label={'Cancel bisect'}
              icon={'cross'}
              onClick={this.terminateBisect}
            >
              Cancel Bisect
            </Button>
          </>
        );
      } else {
        return (
          <Button
            icon="git-branch"
            text="Bisect"
            disabled={appState.isAutoBisecting}
            onClick={appState.toggleBisectDialog}
          />
        );
      }
    }
  },
);
