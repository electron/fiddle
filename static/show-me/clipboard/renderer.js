window.addEventListener('DOMContentLoaded', () => {
  const copyButton = document.querySelector('#copy')
  const pasteButton = document.querySelector('#paste')
  const textarea = document.querySelector('textarea')
  copyButton.onclick = () => {
    window.clipboard.writeText('Hello from Electron!')
  }
  pasteButton.onclick = async () => {
    textarea.value = await window.clipboard.readText()
  }
})
