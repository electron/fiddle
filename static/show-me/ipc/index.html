<!DOCTYPE html>
<html lang="en">
<head>
  <title>IPC Example</title>
</head>
<body>
  <h1>IPC Example</h1>
  <p>Open the developer tools to see the IPC in action.</p>
</body>
</html>
