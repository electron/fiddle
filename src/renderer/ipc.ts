import { EventEmitter } from 'events';

import { ipcRenderer } from 'electron';

import { IpcEvents, ipcRendererEvents } from '../ipc-events';

/**
 * The main purpose of this class is to be the central
 * gathering place for IPC calls the main process sends
 * or listens to.
 *
 * @class IpcManager
 * @extends {EventEmitter}
 */
class IpcRendererManager extends EventEmitter {
  constructor() {
    super();

    ipcRendererEvents.forEach((name) => {
      ipcRenderer.removeAllListeners(name);
      ipcRenderer.on(name, (...args: Array<any>) => this.emit(name, ...args));
    });
  }

  /**
   * Central method to send an IPC message to the main process.
   * Currently very simple, but centralized here so that future
   * refactors will be a lot easier.
   *
   * @param {IpcEvents} channel
   * @param {...Array<any>} args
   * @memberof IpcRendererManager
   */
  public send(channel: IpcEvents, ...args: Array<any>) {
    ipcRenderer.send(channel, ...args);
  }

  public invoke(channel: IpcEvents, ...args: Array<any>) {
    return ipcRenderer.invoke(channel, ...args);
  }
}

export const ipcRendererManager = new IpcRendererManager();
