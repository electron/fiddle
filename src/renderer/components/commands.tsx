import * as React from 'react';

import { Button, ControlGroup } from '@blueprintjs/core';
import { observer } from 'mobx-react';

import { IpcEvents } from '../../ipc-events';
import { ipcRendererManager } from '../ipc';
import { AppState } from '../state';
import { GistActionButton } from './commands-action-button';
import { AddressBar } from './commands-address-bar';
import { BisectHandler } from './commands-bisect';
import { Runner } from './commands-runner';
import { VersionChooser } from './commands-version-chooser';

interface CommandsProps {
  appState: AppState;
}

/**
 * The command bar, containing all the buttons doing
 * all the things
 *
 * @class Commands
 * @extends {React.Component<CommandsProps>}
 */
export const Commands = observer(
  class Commands extends React.Component<CommandsProps> {
    constructor(props: CommandsProps) {
      super(props);
    }

    private handleDoubleClick = (e: React.MouseEvent<HTMLDivElement>) => {
      if (e.target['tagName'] !== 'INPUT') {
        ipcRendererManager.send(IpcEvents.CLICK_TITLEBAR_MAC);
      }
    };

    public render() {
      const { appState } = this.props;
      const { isBisectCommandShowing, title } = appState;

      return (
        <div
          className={
            process.platform === 'darwin' ? 'commands is-mac' : 'commands'
          }
          onDoubleClick={this.handleDoubleClick}
        >
          <div>
            <ControlGroup fill={true} vertical={false}>
              <Button
                icon="cog"
                title="Setting"
                onClick={appState.toggleSettings}
              />
            </ControlGroup>
            <ControlGroup fill={true} vertical={false}>
              <VersionChooser appState={appState} />
              <Runner appState={appState} />
            </ControlGroup>
            {isBisectCommandShowing && (
              <ControlGroup fill={true} vertical={false}>
                <BisectHandler appState={appState} />
              </ControlGroup>
            )}
            <ControlGroup fill={true} vertical={false}>
              <Button
                active={appState.isConsoleShowing}
                icon="console"
                text="Console"
                onClick={appState.toggleConsole}
              />
            </ControlGroup>
          </div>
          {process.platform === 'darwin' ? (
            <div className="title">{title}</div>
          ) : undefined}
          <div>
            <AddressBar appState={appState} />
            <GistActionButton appState={appState} />
          </div>
        </div>
      );
    }
  },
);
