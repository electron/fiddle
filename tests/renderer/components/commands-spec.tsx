import * as React from 'react';

import { Button, ControlGroup } from '@blueprintjs/core';
import { shallow } from 'enzyme';

import { IpcEvents } from '../../../src/ipc-events';
import { Commands } from '../../../src/renderer/components/commands';
import { BisectHandler } from '../../../src/renderer/components/commands-bisect';
import { ipcRendererManager } from '../../../src/renderer/ipc';
import { StateMock } from '../../mocks/mocks';
import { overridePlatform, resetPlatform } from '../../utils';

jest.mock('../../../src/renderer/components/commands-runner', () => ({
  Runner: 'runner',
}));

jest.mock('../../../src/renderer/components/commands-version-chooser', () => ({
  VersionChooser: 'version-chooser',
}));

jest.mock('../../../src/renderer/components/commands-address-bar', () => ({
  AddressBar: 'address-bar',
}));

jest.mock('../../../src/renderer/components/commands-action-button', () => ({
  GistActionButton: 'action-button',
}));

describe('Commands component', () => {
  let store: StateMock;

  beforeEach(() => {
    overridePlatform('linux');
    ({ state: store } = (window as any).ElectronFiddle.app);
  });

  afterEach(() => {
    resetPlatform();
  });

  it('renders when system is darwin', () => {
    overridePlatform('darwin');
    const wrapper = shallow(<Commands appState={store as any} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders when system not is darwin', () => {
    overridePlatform('win32');
    const wrapper = shallow(<Commands appState={store as any} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('can show the bisect command tools', () => {
    store.isBisectCommandShowing = true;
    const wrapper = shallow(<Commands appState={store as any} />);

    expect(wrapper.find(BisectHandler).length).toBe(1);
  });

  it('handleDoubleClick()', () => {
    const spy = jest.spyOn(ipcRendererManager, 'send');

    const wrapper = shallow(<Commands appState={store as any} />);
    const instance = wrapper.instance() as any;

    instance.handleDoubleClick({ target: {} });

    expect(spy).toHaveBeenCalledWith(IpcEvents.CLICK_TITLEBAR_MAC);
    spy.mockRestore();
  });

  it('handleDoubleClick() should not handle input tag', () => {
    const spy = jest.spyOn(ipcRendererManager, 'send');

    const wrapper = shallow(<Commands appState={store as any} />);
    const instance = wrapper.instance() as any;

    instance.handleDoubleClick({ target: { tagName: 'INPUT' } });

    expect(spy).toHaveBeenCalledTimes(0);
    spy.mockRestore();
  });

  it('show setting', () => {
    const wrapper = shallow(<Commands appState={store as any} />);

    wrapper.find(ControlGroup).at(0).find(Button).simulate('click');

    expect(store.toggleSettings).toHaveBeenCalled();
  });
});
