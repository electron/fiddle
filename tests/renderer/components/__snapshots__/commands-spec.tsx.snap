// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands component renders when system is darwin 1`] = `
<div
  className="commands is-mac"
  onDoubleClick={[Function]}
>
  <div>
    <Blueprint3.ControlGroup
      fill={true}
      vertical={false}
    >
      <Blueprint3.Button
        icon="cog"
        onClick={[MockFunction]}
        title="Setting"
      />
    </Blueprint3.ControlGroup>
    <Blueprint3.ControlGroup
      fill={true}
      vertical={false}
    >
      <version-chooser
        appState={"default StateMock"}
      />
      <runner
        appState={"default StateMock"}
      />
    </Blueprint3.ControlGroup>
    <Blueprint3.ControlGroup
      fill={true}
      vertical={false}
    >
      <Blueprint3.Button
        active={true}
        icon="console"
        text="Console"
      />
    </Blueprint3.ControlGroup>
  </div>
  <div
    className="title"
  >
    Electron Fiddle
  </div>
  <div>
    <address-bar
      appState={"default StateMock"}
    />
    <action-button
      appState={"default StateMock"}
    />
  </div>
</div>
`;

exports[`Commands component renders when system not is darwin 1`] = `
<div
  className="commands"
  onDoubleClick={[Function]}
>
  <div>
    <Blueprint3.ControlGroup
      fill={true}
      vertical={false}
    >
      <Blueprint3.Button
        icon="cog"
        onClick={[MockFunction]}
        title="Setting"
      />
    </Blueprint3.ControlGroup>
    <Blueprint3.ControlGroup
      fill={true}
      vertical={false}
    >
      <version-chooser
        appState={"default StateMock"}
      />
      <runner
        appState={"default StateMock"}
      />
    </Blueprint3.ControlGroup>
    <Blueprint3.ControlGroup
      fill={true}
      vertical={false}
    >
      <Blueprint3.Button
        active={true}
        icon="console"
        text="Console"
      />
    </Blueprint3.ControlGroup>
  </div>
  <div>
    <address-bar
      appState={"default StateMock"}
    />
    <action-button
      appState={"default StateMock"}
    />
  </div>
</div>
`;
